NetSuite Documentation Update (#554)
rudderlabs/rudderstack-docs: cloud-extract-sources/netsuite.mdx

# NetSuite

[NetSuite](https://www.netsuite.com/portal/home.shtml) is a popular business management suite that delivers efficient ERP, CRM, and other e-commerce solutions to nearly 25000 customers worldwide.

> This document guides you in setting up Sendgrid as a source in RudderStack. Once configured, RudderStack automatically ingests your specified Sendgrid data, which can then be routed to your RudderStack-supported data warehouse destination.

Reviewer: ```suggestion
This document guides you in setting up NetSuite as a source in RudderStack. Once configured, RudderStack automatically ingests your specified NetSuite data, which can then be routed to your RudderStack-supported data warehouse destination.
```

<div class="warningBlock">

All the Cloud Extract sources support sending data only to a <a href="https://www.rudderstack.com/docs/data-warehouse-integrations/">data warehouse destination</a>.
</div>

## Configuring NetSuite

Before you set up a Netsuite source in RudderStack, there are some RudderStack-specific configurations you need to do in your Netsuite account.

### Whitelisting the IPs

Depending on the setup of your NetSuite organization, you will have to whitelist the following RudderStack IPs:

- 3.216.35.97
- 34.198.90.241
- 54.147.40.62
- 23.20.96.9
- 18.214.35.254

To whitelist the IPs in NetSuite, follow these steps:

1. Sign into your NetSuite account as an administrator.
2. Go to **Setup** > **Company** > **Company Information**.
3. Add a comma-delimited list of the above-mentioned RudderStack IPs in the **Allowed IP addresses** field.
4. Finally, click on **Save**.

### Enabling token-based authentication

> You will also be required to enable token-based authentication for your NetSuite account in order to generate tokens and authenticate the [SuiteTalk API](https://www.netsuite.com/portal/developers/resources/suitetalk-documentation.shtml).

Reviewer: ```suggestion
You will also need to enable token-based authentication for your NetSuite account in order to generate tokens and authenticate the [SuiteTalk API](https://www.netsuite.com/portal/developers/resources/suitetalk-documentation.shtml).
```

To do so, follow these steps:

1. Locate the **Manage Authentication** option in the **Enable Features** page.
2. Enable (check) the **Token-based Authentication** box.
3. Finally, scroll to the bottom and click on **Save**.

<img src="../assets/cloud-extract/netsuite-settings-1.png" alt="Token based Authentication in Netsuite" />

### Creating an integration

Next, create an integration record for RudderStack to uniquely identify RudderStack in your NetSuite account.

<img src="../assets/cloud-extract/netsuite-settings-2.png" alt="New integration in Netsuite" />

Follow these steps to create a new integration:

1. Go to **Setup** > **Integration** > **Manage Integrations** > **New**.
2. In the **New Integration** page, enter the following fields:
    - **Name**: Enter a name for your integration.
    - **State**: This field should be **Enabled**.
3. In the **Authentication** tab, enable (check) **Token-based Authentication**.
4. Finally, click on the **Save** button. The confirmation page will display a **Consumer key/secret** section, as shown:

> <img src="../assets/cloud-extract/netsuite-settings-3.png" alt="Consume key/token in Netsuite" />

Reviewer: ```suggestion
<img src="../assets/cloud-extract-sources/netsuite-settings-3.png" alt="Consume key/token in Netsuite" />
```

<div class="warningBlock">

<strong>Safely store the consumer key and secret</strong>. You will need these credentials while setting up the NetSuite source in RudderStack.
</div>

### Creating a NetSuite role and configuring permissions

To create a NetSuite Token ID and Secret, you first need to create a **Role** for RudderStack and an **Employee** with this specific role.

#### Step 1: Creating a role

To create a new role, follow these steps:

1. Go to **Setup** > **Users/Roles** > **Manage Roles** > **New**.
2. Enter a name for the role in the **Name** field, for example, `RudderStack`.
3. Select **Two-Factor Authentication Required** to **Not Required**.

> <img src="../assets/cloud-extract/netsuite-settings-4.png" alt="2FA Authentication in Netsuite" />

Reviewer: ```suggestion
<img src="../assets/cloud-extract-sources/netsuite-settings-4.png" alt="2FA Authentication in Netsuite" />
```

4. Under the **Permissions** section, go to **Setup** and assign the following permissions:

| Full permissions | View permissions |
| :---- | :----- |
| Web Services | Custom Fields |
| User Access Tokens | Custom Body Fields
| Log In Using Access Tokens | Perform Search (this can be found in the **Lists** section next to **Setup**)
| Deleted Records | - |

> <img src="../assets/cloud-extract/netsuite-settings-5.png" alt="Permissions in Netsuite" />

Reviewer: ```suggestion
<img src="../assets/cloud-extract-sources/netsuite-settings-5.png" alt="Permissions in Netsuite" />
```

For each pipeline resource that you want RudderStack to sync, you should give the relevant permissions for it to the role. These resources are divided among the following sections:

- **Transactions**
- **Reports**

> - **Lists**, and 

Reviewer: ```suggestion
- **Lists**
```

#### Step 2: Creating an employee

To create a new employee, follow these steps:

1. Navigate to **Lists** > **Employees** > **Employees** > **New**.
2. Then, enter the name and email ID of the employee.
3. Under the **Access** section, enable (check) **Give Access** and **Manually assign or change password**. Then add the required password.
4. Finally, add the **Role** which you have created in the previous step and click on **Save** to save the settings.

> <img src="../assets/cloud-extract/netsuite-settings-6.png" alt="Adding an employee in Netsuite" />

Reviewer: ```suggestion
<img src="../assets/cloud-extract-sources/netsuite-settings-6.png" alt="Adding an employee in Netsuite" />
```

#### Step 3: Creating an access token

In this step, you will need to generate the access tokens required for the integration record \(application\) and the user role.

<div class="infoBlock">

To generate the access token, you may have to first log in as this user.
</div>

> <img src="../assets/cloud-extract/netsuite-settings-7.png" alt="Access token settings in Netsuite" />

Reviewer: ```suggestion
<img src="../assets/cloud-extract-sources/netsuite-settings-7.png" alt="Access token settings in Netsuite" />
```

1. Go to **Setup** > **Users/Roles** > **Access Tokens** > **New**.
2. Then, enter the following settings:
    - **Application Name**: Select the integration record you created.
    - **User**: Select the user you created.
    - **Role**: Select the role you created.
    - **Token Name**: Assign a name for the token.
3. Finally, click on the **Save** button. The confirmation page will display a **Token ID and Secret**, as shown:

<div class="warningBlock">

For security reasons, this is the only time these credentials will be displayed. Hence, copy and store the <strong>Token ID</strong> and <strong>Token Secret</strong> safely. You will need these credentials while setting up the NetSuite source in RudderStack.
</div>

> <img src="../assets/cloud-extract/netsuite-settings-8.png" alt="Access token settings in Netsuite" />

Reviewer: ```suggestion
<img src="../assets/cloud-extract-sources/netsuite-settings-8.png" alt="Access token settings in Netsuite" />
```

### Locating your NetSuite Account ID

To locate your NetSuite account ID, go to **Setup** > **Integration** > **Web Services Preferences**.

You will find the account ID in the resulting page, as highlighted below:

> <img src="../assets/cloud-extract/netsuite-settings-9.png" alt="Account ID in Netsuite" />

Reviewer: ```suggestion
<img src="../assets/cloud-extract-sources/netsuite-settings-9.png" alt="Account ID in Netsuite" />
```

<div class="warningBlock">
  
If your Account ID contains a suffix, for example, <code class="inline-code">1234567_SB2</code>, it should be included when you enter the ID during the source configuration in RudderStack.
</div>

## Configuring NetSuite source in RudderStack

To set up NetSuite as a source on the RudderStack dashboard, follow these steps:

1. Log into your [RudderStack dashboard](https://app.rudderstack.com/).

> 2. Go to **Sources** > **New source** > **Cloud Extract** and select **Sendgrid** from the list of sources.

Reviewer: ```suggestion
2. Go to **Sources** > **New source** > **Cloud Extract** and select **NetSuite** from the list of sources.
```

3. Assign a name to your source and click on **Next**.

### Connection settings

Enter the following connection credentials required to set up your NetSuite source in RudderStack:

> <img src="../assets/cloud-extract/netsuite-settings-10.png" alt="Netsuite connection settings" />

Reviewer: ```suggestion
<img src="../assets/cloud-extract-sources/netsuite-settings-10.png" alt="Netsuite connection settings" />
```

* **Account Name** : Enter your NetSuite account name here. You can get it from your NetSuite dashboard.
* **Account ID** : Enter the NetSuite account ID. You can get the account ID by going to your NetSuite dashboard and navigating to **Setup** > **Integration** > **Web Services Preferences**. For more information, refer to the [Locating your NetSuite Account ID](#locating-your-netsuite-account-id) section above.

> * **Consumer Key** : Your NetSuite consumer key goes here.

Reviewer: ```suggestion
* **Consumer Key** : Enter your NetSuite consumer key.
```

> * **Consumer Secret** : Your NetSuite consumer secret goes here.

Reviewer: ```suggestion
* **Consumer Secret** :  Enter your NetSuite consumer secret.
```

<div class="infoBlock">

For more information on getting your NetSuite consumer key and secret, refer to the <a href="#creating-an-integration">Creating an integration</a> section above.
</div>

* **Token ID** : Enter your NetSuite token ID. 

> * **Token Secret** : Your NetSuite token secret needs to be entered here.

Reviewer: ```suggestion
* **Token Secret** : Enter your NetSuite token secret.
```

<div class="infoBlock">

For more information on getting your NetSuite token ID and secret, refer to the <a href="#step-3-creating-an-access-token">Creating an access token</a> section above.
</div>

### Setting the historical sync date

Set the **Historical sync start date** from when you want RudderStack all your historical NetSuite data. Also specify the **RudderStack Restlet URL** and click **Next**, as shown:

> <img src="../assets/cloud-extract/netsuite-settings-11.png" alt="Netsuite connection settings" />

Reviewer: ```suggestion
<img src="../assets/cloud-extract-sources/netsuite-settings-11.png" alt="Netsuite connection settings" />
```

</div>

### Destination settings

The following settings specify how RudderStack sends the data ingested from NetSuite to the connected warehouse destination:

- **Table prefix**: RudderStack uses this prefix to create a table in your data warehouse and loads all your NetSuite data into it.
- **Schedule Settings**: RudderStack gives you three options to ingest the data from NetSuite:
    - **Basic**: Runs the syncs at the specified time interval. 
    - **CRON**: Runs the syncs based on the user-defined CRON expression.
    - **Manual**: You are required to run the syncs manually.

<div class="infoBlock">

For more information on the schedule types, refer to the <a href="https://www.rudderstack.com/docs/cloud-extract-sources/common-settings/">Common Settings</a> guide.
</div>

### Selecting the data to import

Finally, you need to select the NetSuite data that you wish to ingest via RudderStack. You can either select all the data, or choose specific NetSuite data attributes, as per your requirement, as shown:

<div class="warningBlock">

Note that none of the resources are selected by default, so you will have to do the selection manually.
</div>

> <img src="../assets/cloud-extract/netsuite-settings-12.png" alt="Selecting import data" />

Reviewer: ```suggestion
<img src="../assets/cloud-extract-sources/netsuite-settings-12.png" alt="Selecting import data" />
```

#### Importing custom records and saved searches

RudderStack supports two special types of resources that you can ingest from NetSuite, namely **custom records** and **saved searches**.  To import these resources, click on **Add Resource** on the top right, as shown above. You will see the following window:

> <img src="../assets/cloud-extract/netsuite-settings-13.png" alt="Cloning resources" />

Reviewer: ```suggestion
<img src="../assets/cloud-extract-sources/netsuite-settings-13.png" alt="Cloning resources" />
```

<div class="warningBlock">

To get the <strong>RudderStack Restlet URL</strong>, you need to first install <a href="https://js.blendo.co/netsuite/v1/restlet.js">this script</a> in your NetSuite account and paste the generated restlet URL in the RudderStack dashboard.
</div>

To create a new **Script** in your NetSuite account, follow these steps:

1. Go to **Customization** > **Scripting** > **Scripts** > **New**.
2. Then, click on the **+** icon to create a new script. 

> 3. A new window will then pop up asking you to select the downloaded [**script**](https://js.blendo.co/NetSuite/v1/restlet.js).

Reviewer: ```suggestion
3. A new window will then pop up asking you to select the downloaded [script](https://js.blendo.co/NetSuite/v1/restlet.js).
```

4. Once you have attached it, click on **Save**, and then click on **Create Script Record**.

> <img src="../assets/cloud-extract/netsuite-settings-14.png" alt="Creating a new script in NetSuite" />

Reviewer: ```suggestion
<img src="../assets/cloud-extract-sources/netsuite-settings-14.png" alt="Creating a new script in NetSuite" />
```

5. Then, add the **Name** of the script and the **Owner**, that is the **Employee** that you created  in the [Creating an employee](#step-2-creating-an-employee) section above. Then click on **Save**.

> <img src="../assets/cloud-extract/netsuite-settings-15.png" alt="Employee details NetSuite" />

Reviewer: ```suggestion
<img src="../assets/cloud-extract-sources/netsuite-settings-15.png" alt="Employee details NetSuite" />
```

6. To deploy the script, click on **Deployments**, add the script **Title**, and click on **Add**. After the script is deployed, click on **Save**.

> <img src="../assets/cloud-extract/netsuite-settings-16.png" alt="Script deployment NetSuite" />

Reviewer: ```suggestion
<img src="../assets/cloud-extract-sources/netsuite-settings-16.png" alt="Script deployment NetSuite" />
```

7. Finally, click on the title of your deployed script. You should be able to see the **External URL**, which is the **RudderStack Restlet URL** that you need to use in the NetSuite source configuration.

> <img src="../assets/cloud-extract/netsuite-settings-17.png" alt="Restlet URL NetSuite" />

Reviewer: ```suggestion
<img src="../assets/cloud-extract-sources/netsuite-settings-17.png" alt="Restlet URL NetSuite" />
```
